package/sbp_nmea2000_bridge/src/sbp_nmea2000_bridge.cpp

```diff
@@ -0,0 +1,917 @@
+      tm tm_utc_time;
+      // 1900 is the starting year (year 0) in tm_year, so subtract 1900.
+      tm_utc_time.tm_year = sbp_utc_t->year - 1900;
+      // 0 is the starting month in tm_mon, so subtract 1.
+      tm_utc_time.tm_mon = sbp_utc_t->month - 1;
+      tm_utc_time.tm_mday = sbp_utc_t->day;
+      tm_utc_time.tm_hour = sbp_utc_t->hours;
+      tm_utc_time.tm_min = sbp_utc_t->minutes;
+      tm_utc_time.tm_sec = sbp_utc_t->seconds;
+      tm_utc_time.tm_isdst = -1;  // -1 means no idea. mktime() will know.
+      time_t utc_time_since_epoch = mktime(&tm_utc_time);
+
+      constexpr u32 cSecondsInADay = 60 * 60 * 24;
+      *days_since_1970 =
+              static_cast<u16>(utc_time_since_epoch / cSecondsInADay);
+      *seconds_since_midnight =
+              (utc_time_since_epoch % cSecondsInADay) + sbp_utc_t->ns * 1e-9;
+    }
+
+    void piksi_check(int err, const char* format, ...) {
+      if (err != 0) {
+        va_list ap;
+        va_start(ap, format);
+        piksi_log(LOG_ERR, format, ap);
+        va_end(ap);
+        if(debug) {
+          printf(format, ap);
+        }
+        exit(EXIT_FAILURE);
+      }
+    }
+
+    int callback_can_debug(zloop_t *loop, zmq_pollitem_t *item,
+                           void *interface_name_void) {
+      UNUSED(loop);
+      auto interface_name = reinterpret_cast<const char*>(interface_name_void);
+
+      canfd_frame frame;
+      ssize_t bytes_read = recvfrom(item->fd, &frame,sizeof(frame),
+              /*flags=*/0,
+              /*src_addr=*/NULL,
+              /*addrlen=*/NULL);
+      piksi_check(bytes_read < 0, "Could not read interface %s.",
+                  interface_name);
+
+      std::cout << "Read " << bytes_read << " bytes from " << interface_name
+                <<" ID: " << std::setfill('0') << std::setw(4) << std::hex
+                << frame.can_id
+                << " Data: ";
+      for (u32 i = 0; i < frame.len; ++i) {
+        std::cout << std::setfill('0') << std::setw(2) << std::hex
+                  << static_cast<u16>(frame.data[i]);
+      }
+      std::cout << std::endl;
+
+      return 0;
+    }
+
+    // PGN 126992
+    void callback_sbp_utc_time(u16 sender_id, u8 len, u8 msg[], void *context) {
+      UNUSED(sender_id);
+      UNUSED(len);
+      UNUSED(context);
+      d << __FUNCTION__ << "\n";
+
+      auto sbp_utc_time = reinterpret_cast<msg_utc_time_t*>(msg);
+      u16 days_since_1970;
+      double seconds_since_midnight;
+      calculate_days_and_seconds_since_1970(sbp_utc_time,
+                                            &days_since_1970,
+                                            &seconds_since_midnight);
+      d << "\tTOW: " << sbp_utc_time->tow << "\n"
+        << "\tDays since 1970: " << days_since_1970 << "\n"
+        << "\tSeconds since midnight: " << seconds_since_midnight << "\n";
+
+      bool is_valid = is_flags_valid(sbp_utc_time->flags, /*bits=*/3);
```

Comment:
This seems to be checking against a bitmask (0x7?) -- what bits are we looking for?  That is, why would the first 3 bits of the flags on the message dictate whether we look in `tow_sequence_id_cache`?  Then, can we add a comment like "Since SBP UTC time is not valid, send sentinel values."  Also, does it make sense to send a message at all in this case?  Maybe just not send anything?  (The same pattern seems to be replicated in all of the callback_* functions).